File: framework-typescript/src/speaker.ts
// speaker.ts — the MCP STDIO SPEAKER, written here rather than taken from an SDK.
//
// THE DECISION AND ITS MEASUREMENT. This package speaks newline-delimited
// JSON-RPC on stdin and stdout itself and takes NO MCP SDK dependency. The SDK
// route was measured and superseded: against the collector contract's input
// schema advertised VERBATIM, the official TypeScript SDK's low-level server ran
// the handler and returned success for arguments `{}` (the contract's own
// required `id` unsatisfied), for a wrong-typed `params` and for a wrong-typed
// `id`, and it passed a result violating the advertised output schema straight
// through. It validates neither side, so its contribution is framing alone — and
// the framing is the four messages below. Its cost is 93 node_modules entries and
// 116 lockfile entries against 8 and 8 here, in a lockfile this repository
// publishes and leak-scans under a policy that fixes every dependency alert by a
// deliberate pull request. Nothing here is on a hot path; the argument is alert
// surface, and disk is incidental. If you are reading this because you need a
// protocol feature, ADD IT HERE — reaching for the SDK restores the whole tree.
//
// STDOUT IS THE PROTOCOL STREAM. Every byte this module writes to stdout is a
// JSON-RPC message followed by a newline. A stray `console.log` anywhere in a
// collector, a dependency or this package corrupts the framing and surfaces to an
// operator as an opaque handshake failure. Diagnostics go to stderr, where
// `console.error` already writes.
//
// THE CONSUMER SURFACE IS FOUR MESSAGES, source-read on the client's dial path:
// initialize, notifications/initialized, tools/list, tools/call. MCP carries
// more (ping, cancellation, progress, logging), and a request naming a method
// this speaker does not implement is answered with a JSON-RPC "method not found"
// rather than crashing the session; a notification naming one is ignored.

import { createInterface } from "node:readline";

import type { JsonSchema } from "./framework.js";

/**
 * The protocol revisions this speaker can speak, newest first. Every one of them
 * carries the tool `outputSchema` declaration and the `structuredContent` result
 * the collector contract depends on; the collector contract's own floor is
 * 2025-06-18, which is why that is what an unrecognised request is answered with.
 */
export const SUPPORTED_PROTOCOL_VERSIONS: readonly string[] = [
  "2026-07-28",
  "2025-11-25",
  "2025-06-18",
];

/**
 * The revision an `initialize` naming an unsupported one is answered with. It is
 * the collector contract's feature floor rather than the newest revision: a
 * caller that asked for something this speaker does not know is told the oldest
 * revision that still carries everything the contract needs.
 */
export const PREFERRED_PROTOCOL_VERSION = "2025-06-18";

/** One tool as `tools/list` carries it. */
export interface ToolDefinition {
  name: string;
  description?: string;
  inputSchema: JsonSchema;
  /** Omitted entirely when absent, which is a shape a conformance stub needs. */
  outputSchema?: JsonSchema;
}

/** One content block of a tool result. Only text is used by this package. */
export interface TextContent {
  type: "text";
  text: string;
}

/** A `tools/call` result, in the shape the wire carries it. */
export interface CallToolResult {
  content?: TextContent[];
  structuredContent?: unknown;
  isError?: boolean;
}

/** What a speaker serves: a tool listing and one handler for every call. */
export interface SpeakerDefinition {
  serverName: string;
  serverVersion: string;
  tools: ToolDefinition[];
  call(name: string, args: Record<string, unknown>): Promise<CallToolResult> | CallToolResult;
}

/**
 * The transport a speaker reads and writes. It exists so the session can be
 * driven in-process by a test with no child process and no pipes, which is what
 * lets the protocol arms be asserted directly rather than inferred from a
 * client's behaviour.
 */
export interface SpeakerIO {
  /** Called once per outbound message, with the message already newline-terminated. */
  write(line: string): void;
}

/**
 * One live MCP session over a line-oriented transport. Feed it inbound lines with
 * {@link Session.handleLine}; it writes every response through the
 * {@link SpeakerIO} it was built with.
 */
export class Session {
  #negotiated: string | undefined;

  constructor(
    private readonly def: SpeakerDefinition,
    private readonly io: SpeakerIO,
  ) {}

  /**
   * Handles one inbound line. A line that is not JSON, and a JSON value that is
   * not a JSON-RPC message, are answered with an error or ignored; neither ends
   * the session. Resolves once any response has been written.
   */
  async handleLine(line: string): Promise<void> {
    if (line.trim() === "") return;

    let message: unknown;
    try {
      message = JSON.parse(line);
    } catch {
      // A LINE THAT DID NOT PARSE HAS NO ID TO ANSWER, so the response carries
      // the null id the JSON-RPC spec reserves for exactly this. It is answered
      // rather than ignored because a caller waiting on a request it mangled
      // would otherwise wait forever.
      this.#send({ jsonrpc: "2.0", id: null, error: { code: -32700, message: "parse error" } });
      return;
    }
    if (typeof message !== "object" || message === null || Array.isArray(message)) {
      // A scalar, an array or null carries no id, so there is nothing to answer.
      return;
    }
    const request = message as Record<string, unknown>;
    const id = request["id"];
    const method = request["method"];
    const hasId = id !== undefined && id !== null;

    if (typeof method !== "string") {
      if (hasId) {
        this.#send({
          jsonrpc: "2.0",
          id: id as string | number,
          error: { code: -32600, message: "invalid request: no method" },
        });
      }
      return;
    }

    const params = request["params"];
    switch (method) {
      case "initialize": {
        const requested =
          typeof params === "object" && params !== null
            ? (params as Record<string, unknown>)["protocolVersion"]
            : undefined;
        this.#negotiated = negotiateProtocolVersion(requested);
        if (!hasId) return;
        this.#send({
          jsonrpc: "2.0",
          id: id as string | number,
          result: {
            protocolVersion: this.#negotiated,
            capabilities: { tools: {} },
            serverInfo: { name: this.def.serverName, version: this.def.serverVersion },
          },
        });
        return;
      }
      case "notifications/initialized":
        return;
      case "tools/list": {
        if (!hasId) return;
        this.#send({
          jsonrpc: "2.0",
          id: id as string | number,
          result: { tools: this.def.tools.map(toolWire) },
        });
        return;
      }
      case "tools/call": {
        if (!hasId) return;
        const call = typeof params === "object" && params !== null ? (params as Record<string, unknown>) : {};
        const name = typeof call["name"] === "string" ? (call["name"] as string) : "";
        const args =
          typeof call["arguments"] === "object" &&
          call["arguments"] !== null &&
          !Array.isArray(call["arguments"])
            ? (call["arguments"] as Record<string, unknown>)
            : {};
        const result = await this.#invoke(name, args);
        this.#send({ jsonrpc: "2.0", id: id as string | number, result: resultWire(result) });
        return;
      }
      default: {
        // A NOTIFICATION NAMING AN UNKNOWN METHOD IS IGNORED; a REQUEST is
        // answered. MCP carries more methods than the four above, and neither
        // crashing nor hanging is an answer.
        if (!hasId) return;
        this.#send({
          jsonrpc: "2.0",
          id: id as string | number,
          error: { code: -32601, message: `method not found: ${method}` },
        });
      }
    }
  }

  /** The negotiated protocol version, or undefined before `initialize`. */
  negotiatedVersion(): string | undefined {
    return this.#negotiated;
  }

  /**
   * Runs one tool call, turning both refusals a call can meet into a RESULT with
   * isError rather than a JSON-RPC error: the client reads isError as a refused
   * collect that writes nothing, and a protocol-level error is a statement about
   * the request rather than about the walk.
   */
  async #invoke(name: string, args: Record<string, unknown>): Promise<CallToolResult> {
    if (!this.def.tools.some((t) => t.name === name)) {
      // A CALL NAMING A TOOL THIS SPEAKER DOES NOT LIST is refused here, so no
      // handler can be reached under a name it never advertised. The client
      // selects by name from the listing, so a stale entry hits this arm.
      return toolError(
        `this provider serves no tool named "${name}"; it serves ` +
          this.def.tools.map((t) => `"${t.name}"`).join(", "),
      );
    }
    try {
      return await this.def.call(name, args);
    } catch (err) {
      return toolError(messageOf(err));
    }
  }

  #send(message: unknown): void {
    // ONE MESSAGE, ONE LINE. JSON.stringify emits no interior newline for any
    // value, so the framing is the terminator alone.
    this.io.write(JSON.stringify(message) + "\n");
  }
}

/** One tool as the wire carries it, with an absent output schema OMITTED. */
function toolWire(tool: ToolDefinition): Record<string, unknown> {
  const out: Record<string, unknown> = { name: tool.name, inputSchema: tool.inputSchema };
  if (tool.description !== undefined) out["description"] = tool.description;
  // AN ABSENT SCHEMA IS AN ABSENT KEY, never a null one: the client's gate reads
  // "no output schema" from the key's absence, and `null` and a missing key are
  // the same fact to it only by luck.
  if (tool.outputSchema !== undefined) out["outputSchema"] = tool.outputSchema;
  return out;
}

/** One call result as the wire carries it, with every absent field omitted. */
function resultWire(result: CallToolResult): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  if (result.content !== undefined) out["content"] = result.content;
  if (result.structuredContent !== undefined) out["structuredContent"] = result.structuredContent;
  if (result.isError !== undefined) out["isError"] = result.isError;
  return out;
}

/** A tool result carrying a failure as text, with no structured content at all. */
function toolError(text: string): CallToolResult {
  return { isError: true, content: [{ type: "text", text }] };
}

/** The message of a thrown value, whatever kind of value it was. */
function messageOf(err: unknown): string {
  return err instanceof Error ? err.message : String(err);
}

/**
 * Chooses the protocol version an `initialize` naming `requested` is answered
 * with: the requested revision when this speaker supports it, and
 * {@link PREFERRED_PROTOCOL_VERSION} otherwise. It never answers a revision this
 * speaker cannot speak.
 */
export function negotiateProtocolVersion(requested: unknown): string {
  if (typeof requested === "string" && SUPPORTED_PROTOCOL_VERSIONS.includes(requested)) {
    return requested;
  }
  return PREFERRED_PROTOCOL_VERSION;
}

/**
 * Serves one speaker definition over stdin and stdout until stdin closes. This is
 * the entry point a collector binary reaches through `serveStdio`, and the one a
 * conformance stub reaches directly.
 */
export function serveSpeakerOverStdio(def: SpeakerDefinition): Promise<void> {
  const session = new Session(def, {
    write(line: string) {
      process.stdout.write(line);
    },
  });
  const rl = createInterface({ input: process.stdin, crlfDelay: Infinity });
  // LINES ARE HANDLED IN ORDER even though the handler is async: readline emits
  // synchronously, so without this chain a slow call and a fast one would answer
  // out of order and a caller matching responses by arrival would be wrong.
  let queue: Promise<void> = Promise.resolve();
  return new Promise<void>((resolve, reject) => {
    rl.on("line", (line: string) => {
      queue = queue.then(() => session.handleLine(line)).catch(reject);
    });
    rl.on("close", () => {
      queue.then(resolve).catch(reject);
    });
  });
}

File: framework-python/knowledge_collector/speaker.py
# SPDX-License-Identifier: Apache-2.0

"""speaker.py — the MCP STDIO SPEAKER, hand-rolled on the standard library.

WHY THIS FILE EXISTS AT ALL, because it is the one part of the port with no
counterpart in the Go framework. The Go framework hands its server to an SDK and
the SDK owns the framing, the handshake and the malformed-frame refusals. This
port takes NO RUNTIME DEPENDENCY, so those obligations are its own, and they are
tested as its own: the protocol-version floor, the initialized notification, the
unbounded newline framing, and four malformed-frame arms.

THE SURFACE IS FOUR MESSAGES AND IT IS FIXED BY A CHECKED-IN CONTRACT. The
knowledge client dials a collector, lists its tools, and calls one:

    initialize                 -> the handshake, answering this speaker's version
    notifications/initialized  -> a NOTIFICATION; it is answered with NOTHING
    tools/list                 -> the advertised input and output schemas
    tools/call                 -> the walk

`ping` is answered too, because it is part of the base protocol and costs one
branch. Anything else is method-not-found, by name.

STDOUT IS THE PROTOCOL STREAM. A collector that prints to stdout corrupts the
JSON-RPC framing and surfaces to an operator as an opaque handshake failure;
every diagnostic this package writes goes to stderr, and the port's suite pins
that with a real child process rather than with a reading of this comment.

THE FRAMING IS NEWLINE-DELIMITED JSON WITH NO LENGTH BOUND IN EITHER DIRECTION.
Collector traffic carries no size cap: a result crosses 64 MiB on a real account
long before it is interesting, and a declared foreign-context block crosses 4 MiB
on an ordinary one. `readline()` on a buffered binary stream is unbounded, and one
result is written as ONE newline-terminated line with no chunking of its own. A
fixed-size read or a line-length cap here is the defect the port's own suite
exists to catch, because it fails nowhere else.
"""

import json
import sys

__all__ = [
    "PROTOCOL_VERSION",
    "JSONRPC_VERSION",
    "PARSE_ERROR",
    "INVALID_REQUEST",
    "METHOD_NOT_FOUND",
    "INVALID_PARAMS",
    "INTERNAL_ERROR",
    "ToolDefinition",
    "Speaker",
]

# The protocol revision this speaker ANSWERS. It is a constant of the port, not
# an echo of what the caller asked for: the client requests its own newest
# revision and accepts a range below it, so a speaker that mirrored the request
# would claim to support whatever it was handed.
PROTOCOL_VERSION = "2025-06-18"

JSONRPC_VERSION = "2.0"

# The JSON-RPC 2.0 error codes this speaker emits.
PARSE_ERROR = -32700
INVALID_REQUEST = -32600
METHOD_NOT_FOUND = -32601
INVALID_PARAMS = -32602
INTERNAL_ERROR = -32603


class ToolDefinition:
    """One tool this speaker serves: what `tools/list` advertises and what
    `tools/call` dispatches to.

    `handler` takes the call's decoded `arguments` object and returns the tool
    result document (a dict with any of `content`, `structuredContent`,
    `isError`). It raises to signal a walk failure, which becomes an isError
    result rather than a JSON-RPC error, because the client treats a tool error
    as a refused collect that writes nothing and a transport error as a broken
    provider.
    """

    __slots__ = ("name", "description", "input_schema", "output_schema", "handler")

    def __init__(self, name, description, input_schema, output_schema, handler):
        self.name = name
        self.description = description
        self.input_schema = input_schema
        self.output_schema = output_schema
        self.handler = handler

    def listing(self):
        """The tool as `tools/list` carries it.

        AN ABSENT OUTPUT SCHEMA IS OMITTED RATHER THAN SENT AS null. A provider
        that advertises no output schema is a real shape the client refuses by
        name, and it is reachable here by passing None; sending `"outputSchema":
        null` would be a third thing that is neither.
        """
        out = {"name": self.name, "description": self.description, "inputSchema": self.input_schema}
        if self.output_schema is not None:
            out["outputSchema"] = self.output_schema
        return out


class Speaker:
    """Serves a set of tools over MCP on newline-delimited JSON.

    The streams are parameters rather than module globals so the port's own suite
    can drive a speaker in-process over byte buffers AND as a real child process,
    and get the same code path both times.
    """

    def __init__(self, server_name, server_version, tools, stdin=None, stdout=None, stderr=None):
        self.server_name = server_name
        self.server_version = server_version
        self.tools = {tool.name: tool for tool in tools}
        self._stdin = stdin if stdin is not None else sys.stdin.buffer
        self._stdout = stdout if stdout is not None else sys.stdout.buffer
        self._stderr = stderr if stderr is not None else sys.stderr

    # ---------------------------------------------------------------- the loop --

    def serve(self):
        """Read frames until end of input, answering each. Returns on EOF."""
        while True:
            line = self._stdin.readline()
            if not line:
                return
            self.handle_frame(line)

    def handle_frame(self, line):
        """Answer ONE newline-terminated frame.

        SIX MALFORMED-INPUT ARMS, each its own branch so a mutation that replaces
        one with a skip turns exactly one test red. The repository's hard
        invariant is that bad input always errors with no silent coercion,
        default or degrade, and this reader is the one place in this package that
        reads untrusted bytes.

        THE ARMS ARE AT TWO LEVELS AND THE SECOND LEVEL IS THE ONE THAT BITES.
        Four are about the FRAME -- not JSON, not a request object, an unknown
        method, no id. The fifth is about a MEMBER of a well-formed frame: a
        request object whose `params` is an array, a string or a number. Measured
        on the shipped speaker before this arm existed, each of those raised an
        AttributeError out of serve() and ENDED THE SESSION with zero bytes on
        stdout -- no error for the offending frame and no answer to the
        well-formed frame behind it. A refusal is a RESPONSE, not a disconnect,
        and that property was false for exactly the classes no row drove.
        """
        text = line.decode("utf-8", errors="replace").strip()

        # (1) A frame that is not JSON at all -- including an empty one, which
        # carries no message and is a framing fault rather than a quiet nothing.
        try:
            message = json.loads(text) if text else None
        except ValueError as exc:
            self.log("refusing a frame that is not JSON: %s" % exc)
            self._write(self._error_envelope(None, PARSE_ERROR, "the frame is not JSON: %s" % exc))
            return
        if message is None:
            self.log("refusing an empty frame; a newline carries no JSON-RPC message")
            self._write(self._error_envelope(None, PARSE_ERROR, "the frame carries no JSON-RPC message"))
            return

        # (2) JSON that is not a JSON-RPC request object.
        if not isinstance(message, dict) or "method" not in message or not isinstance(message.get("method"), str):
            self.log("refusing a frame that is not a JSON-RPC request object")
            self._write(
                self._error_envelope(
                    message.get("id") if isinstance(message, dict) else None,
                    INVALID_REQUEST,
                    "the frame is not a JSON-RPC request object; it declares no string `method`",
                )
            )
            return

        method = message["method"]
        has_id = "id" in message and message["id"] is not None

        # (3) A `params` MEMBER that is not an object. JSON-RPC allows params to
        # be by-position (an array) or by-name (an object); MCP is by-name only,
        # so an array, a string or a number here is a caller that cannot be
        # served. It is refused with INVALID_PARAMS naming what arrived, rather
        # than reaching a `.get` on a list and taking the session down with it.
        #
        # ABSENT AND null BOTH MEAN "no params" and are the empty object: the
        # member MAY be omitted, and a caller that sends null has omitted it in a
        # second spelling rather than sent a malformed one.
        params = message.get("params")
        if params is None:
            params = {}
        elif not isinstance(params, dict):
            self.log("refusing a frame whose params is a %s" % type(params).__name__)
            self._write(
                self._error_envelope(
                    message.get("id"),
                    INVALID_PARAMS,
                    "the request's `params` is a %s; this protocol passes parameters BY NAME, so params must be an object"
                    % type(params).__name__,
                )
            )
            return

        # (4) A frame with no id is a NOTIFICATION: it is answered with nothing,
        # which is the protocol rather than a skip, and it is logged so a silent
        # drop is not what an operator sees.
        if not has_id:
            self._handle_notification(method)
            return

        request_id = message["id"]
        if method == "initialize":
            self._write(self._result_envelope(request_id, self._initialize_result()))
            return
        if method == "ping":
            self._write(self._result_envelope(request_id, {}))
            return
        if method == "tools/list":
            self._write(self._result_envelope(request_id, {"tools": [t.listing() for t in self.tools.values()]}))
            return
        if method == "tools/call":
            self._write(self._result_envelope(request_id, self._call_tool(params)))
            return

        # (5) A method this speaker does not serve, named in the refusal.
        self.log("refusing an unknown method %r" % method)
        self._write(self._error_envelope(request_id, METHOD_NOT_FOUND, "unknown method %r" % method))

    def _handle_notification(self, method):
        if method == "notifications/initialized":
            self.log("handshake complete")
            return
        self.log("ignoring the notification %r; a notification carries no id and is answered with nothing" % method)

    # ------------------------------------------------------------- the answers --

    def _initialize_result(self):
        """The handshake answer.

        `serverInfo.name` IS THE TOOL NAME on a single-tool collector, which is
        what the framework's entry point passes in: the client logs it, and a name
        that matched nothing an operator configured would be the one diagnostic
        that helps them least.
        """
        return {
            "protocolVersion": PROTOCOL_VERSION,
            "capabilities": {"tools": {}},
            "serverInfo": {"name": self.server_name, "version": self.server_version},
        }

    def _call_tool(self, params):
        """Dispatch one `tools/call`, returning the CallToolResult document."""
        name = params.get("name")

        # (6) A tool NAME that is not a string. It is refused before it reaches
        # the tool table, because an unhashable value -- an array, an object --
        # raises a TypeError out of a dict lookup and takes the session with it,
        # which is the same class as the params arm above one level further in.
        if not isinstance(name, str):
            self.log("refusing a tools/call whose name is a %s" % type(name).__name__)
            return self._tool_error(
                "the tools/call `name` is a %s; it names the tool to call and must be a string" % type(name).__name__
            )

        tool = self.tools.get(name)
        if tool is None:
            return self._tool_error("this provider serves no tool named %r; it serves %s" % (name, ", ".join(sorted(self.tools))))
        arguments = params.get("arguments")
        if arguments is None:
            arguments = {}
        try:
            result = tool.handler(arguments)
        except Exception as exc:  # noqa: BLE001 -- a walk failure is a TOOL error
            # A walk failure is a TOOL error, never an empty successful result:
            # the client treats an isError result as a refused collect that writes
            # nothing, while an empty complete result would instead assert a
            # successful walk that found nothing.
            self.log("the tool %r failed: %s" % (name, exc))
            return self._tool_error(str(exc))
        return result

    @staticmethod
    def _tool_error(text):
        return {"content": [{"type": "text", "text": text}], "isError": True}

    # ------------------------------------------------------------ the framing --

    def _result_envelope(self, request_id, result):
        return {"jsonrpc": JSONRPC_VERSION, "id": request_id, "result": result}

    def _error_envelope(self, request_id, code, message):
        return {"jsonrpc": JSONRPC_VERSION, "id": request_id, "error": {"code": code, "message": message}}

    def _write(self, document):
        """Write ONE document as ONE newline-terminated line.

        No chunking of its own and no length bound: a 68 MiB result is one line on
        the wire, exactly as the client's reader expects it.
        """
        payload = json.dumps(document, separators=(",", ":")).encode("utf-8")
        self._stdout.write(payload)
        self._stdout.write(b"\n")
        self._stdout.flush()

    def log(self, message):
        """Every diagnostic goes to STDERR. stdout is the protocol stream."""
        print("%s: %s" % (self.server_name, message), file=self._stderr)
